File: Packages/Sources/RxCodeCore/Secrets/SecretsCrypto.swift
import CryptoKit
import Foundation

/// Pure CryptoKit port of github-pm's `lib/secrets/crypto.ts`. Keeping this
/// byte-for-byte interoperable with the web app is the whole point: a secret
/// enrolled / encrypted in the browser must decrypt here and vice-versa.
///
/// The scheme:
/// - A 32-byte **PRF output** is extracted from the user's passkey (done in the
///   app layer via `AuthenticationServices`; this type never touches passkeys).
/// - The **KEK** (key-encryption key) is `HKDF-SHA256(prf, salt: ∅, info: "…/kek")`.
/// - Each user has an **ECDH P-256 keypair**; the private key is exported as a
///   JWK, AES-GCM-wrapped under the KEK, and stored server-side.
/// - Each environment has a random 32-byte **DEK**; the owner wraps it under
///   their KEK (`wrapMode: "kek"`), and it can also be sealed to another user's
///   public key (`wrapMode: "hpke"`).
/// - Files are AES-GCM encrypted under the DEK.
///
/// AES-GCM layout matches WebCrypto exactly: the `iv` (12 bytes) is stored
/// separately, and the `ciphertext` field is `rawCiphertext ‖ 16-byte tag`
/// (WebCrypto appends the tag; CryptoKit keeps them separate, so we concatenate
/// on the way out and split on the way in).
public enum SecretsCrypto {

    // MARK: - Constants (must match github-pm/lib/secrets/constants.ts)

    /// `github-pm-secrets-v1-prf-salt!!!` — the load-bearing PRF salt. Changing
    /// it invalidates every wrapped private key. 32 bytes.
    public static let prfSalt = Data("github-pm-secrets-v1-prf-salt!!!".utf8)

    /// HKDF `info` for deriving the KEK from PRF output.
    public static let hkdfInfoKEK = Data("github-pm-secrets-v1/kek".utf8)

    /// HKDF `info` for deriving an AES-GCM key from an ECDH shared secret (HPKE).
    public static let hkdfInfoHPKE = Data("github-pm-secrets-v1/hpke".utf8)

    /// WebAuthn relying-party identifier the passkeys are scoped to.
    public static let webAuthnRPID = "rxlab.app"

    public enum CryptoError: Error, LocalizedError {
        case badBase64(String)
        case ciphertextTooShort
        case invalidJWK
        case prfUnavailable

        public var errorDescription: String? {
            switch self {
            case .badBase64(let field): return "Malformed base64 in \(field)."
            case .ciphertextTooShort: return "Ciphertext is too short to contain an auth tag."
            case .invalidJWK: return "Stored private key is not a valid P-256 JWK."
            case .prfUnavailable: return "This passkey does not support the PRF extension."
            }
        }
    }

    // MARK: - KEK

    /// `HKDF-SHA256(prfOutput, salt: ∅, info: "github-pm-secrets-v1/kek")` → 256-bit AES key.
    public static func deriveKEK(prfOutput: Data) -> SymmetricKey {
        HKDF<SHA256>.deriveKey(
            inputKeyMaterial: SymmetricKey(data: prfOutput),
            salt: Data(),
            info: hkdfInfoKEK,
            outputByteCount: 32
        )
    }

    // MARK: - AES-GCM (WebCrypto-compatible serialization)

    /// Encrypts `plaintext`, returning base64 `ciphertext` (`raw ‖ tag`) + base64 `iv`.
    public static func aesGcmEncrypt(
        key: SymmetricKey,
        plaintext: Data
    ) throws -> (ciphertext: String, iv: String) {
        let nonce = AES.GCM.Nonce()
        let sealed = try AES.GCM.seal(plaintext, using: key, nonce: nonce)
        var combined = Data(sealed.ciphertext)
        combined.append(sealed.tag)
        return (combined.base64EncodedString(), Data(nonce).base64EncodedString())
    }

    /// Decrypts a WebCrypto-style `{ciphertext: raw‖tag, iv}` pair.
    public static func aesGcmDecrypt(
        key: SymmetricKey,
        ciphertextB64: String,
        ivB64: String
    ) throws -> Data {
        guard let ctTag = Data(base64Encoded: ciphertextB64) else {
            throw CryptoError.badBase64("ciphertext")
        }
        guard let iv = Data(base64Encoded: ivB64) else {
            throw CryptoError.badBase64("iv")
        }
        guard ctTag.count >= 16 else { throw CryptoError.ciphertextTooShort }
        let tag = Data(ctTag.suffix(16))
        let ct = Data(ctTag.prefix(ctTag.count - 16))
        let box = try AES.GCM.SealedBox(
            nonce: try AES.GCM.Nonce(data: iv),
            ciphertext: ct,
            tag: tag
        )
        return try AES.GCM.open(box, using: key)
    }

    // MARK: - User keypair (ECDH P-256)

    public struct UserKeypair {
        /// Base64 of the raw uncompressed public point (`0x04 ‖ X ‖ Y`, 65 bytes).
        public let publicKeyB64: String
        /// The private key, ready to wrap.
        public let privateKey: P256.KeyAgreement.PrivateKey
    }

    public static func generateUserKeypair() -> UserKeypair {
        let pk = P256.KeyAgreement.PrivateKey()
        return UserKeypair(
            publicKeyB64: pk.publicKey.x963Representation.base64EncodedString(),
            privateKey: pk
        )
    }

    public static func importPublicKey(b64: String) throws -> P256.KeyAgreement.PublicKey {
        guard let data = Data(base64Encoded: b64) else { throw CryptoError.badBase64("publicKey") }
        return try P256.KeyAgreement.PublicKey(x963Representation: data)
    }

    /// Serializes the private key as a WebCrypto-importable JWK (`kty/crv/x/y/d`).
    public static func privateKeyJWK(_ pk: P256.KeyAgreement.PrivateKey) -> [String: Any] {
        let pub = pk.publicKey.x963Representation // 0x04 ‖ X(32) ‖ Y(32)
        let x = pub.subdata(in: 1..<33)
        let y = pub.subdata(in: 33..<65)
        let d = pk.rawRepresentation // 32-byte scalar
        return [
            "kty": "EC",
            "crv": "P-256",
            "x": base64urlEncode(x),
            "y": base64urlEncode(y),
            "d": base64urlEncode(d),
            "ext": true,
        ]
    }

    public static func importPrivateKeyJWK(_ jwk: [String: Any]) throws -> P256.KeyAgreement.PrivateKey {
        guard
            let xS = jwk["x"] as? String, let x = base64urlDecode(xS),
            let yS = jwk["y"] as? String, let y = base64urlDecode(yS),
            let dS = jwk["d"] as? String, let d = base64urlDecode(dS),
            x.count == 32, y.count == 32, d.count == 32
        else { throw CryptoError.invalidJWK }
        var x963 = Data([0x04])
        x963.append(x)
        x963.append(y)
        x963.append(d)
        return try P256.KeyAgreement.PrivateKey(x963Representation: x963)
    }

    /// AES-GCM-wraps the private key (as JWK JSON) under the KEK.
    public static func wrapPrivateKey(
        _ pk: P256.KeyAgreement.PrivateKey,
        kek: SymmetricKey
    ) throws -> (wrappedPrivateKey: String, wrapIv: String) {
        let json = try JSONSerialization.data(
            withJSONObject: privateKeyJWK(pk),
            options: [.sortedKeys]
        )
        let enc = try aesGcmEncrypt(key: kek, plaintext: json)
        return (wrappedPrivateKey: enc.ciphertext, wrapIv: enc.iv)
    }

    public static func unwrapPrivateKey(
        wrappedPrivateKey: String,
        wrapIv: String,
        kek: SymmetricKey
    ) throws -> P256.KeyAgreement.PrivateKey {
        let json = try aesGcmDecrypt(key: kek, ciphertextB64: wrappedPrivateKey, ivB64: wrapIv)
        guard let jwk = try JSONSerialization.jsonObject(with: json) as? [String: Any] else {
            throw CryptoError.invalidJWK
        }
        return try importPrivateKeyJWK(jwk)
    }

    // MARK: - DEK (per-environment data key)

    /// Generates a fresh 32-byte DEK, returning both the key and its raw bytes
    /// (the raw bytes are what get wrapped under a KEK / public key).
    public static func generateDEK() -> (key: SymmetricKey, rawBytes: Data) {
        let key = SymmetricKey(size: .bits256)
        let raw = key.withUnsafeBytes { Data($0) }
        return (key, raw)
    }

    public static func importDEK(rawBytes: Data) -> SymmetricKey {
        SymmetricKey(data: rawBytes)
    }

    // MARK: - HPKE-style sealing (ECDH → HKDF → AES-GCM)

    private static func hpkeKey(shared: SharedSecret) -> SymmetricKey {
        shared.hkdfDerivedSymmetricKey(
            using: SHA256.self,
            salt: Data(),
            sharedInfo: hkdfInfoHPKE,
            outputByteCount: 32
        )
    }

    /// Seals `plaintext` to `recipientPublicKey`, returning the ciphertext, iv
    /// and the ephemeral public key (base64 raw point) needed to unseal.
    public static func sealToPublicKey(
        plaintext: Data,
        recipientPublicKey: P256.KeyAgreement.PublicKey
    ) throws -> (ciphertext: String, wrapIv: String, ephemeralPublicKey: String) {
        let ephemeral = P256.KeyAgreement.PrivateKey()
        let shared = try ephemeral.sharedSecretFromKeyAgreement(with: recipientPublicKey)
        let aesKey = hpkeKey(shared: shared)
        let enc = try aesGcmEncrypt(key: aesKey, plaintext: plaintext)
        return (
            enc.ciphertext,
            enc.iv,
            ephemeral.publicKey.x963Representation.base64EncodedString()
        )
    }

    public static func unsealWithPrivateKey(
        ciphertext: String,
        ivB64: String,
        ephemeralPublicKeyB64: String,
        recipientPrivateKey: P256.KeyAgreement.PrivateKey
    ) throws -> Data {
        let ephemeralPub = try importPublicKey(b64: ephemeralPublicKeyB64)
        let shared = try recipientPrivateKey.sharedSecretFromKeyAgreement(with: ephemeralPub)
        return try aesGcmDecrypt(
            key: hpkeKey(shared: shared),
            ciphertextB64: ciphertext,
            ivB64: ivB64
        )
    }

    // MARK: - File-level helpers

    public static func encryptFile(
        plaintext: String,
        dek: SymmetricKey
    ) throws -> (ciphertext: String, iv: String, size: Int) {
        let bytes = Data(plaintext.utf8)
        let enc = try aesGcmEncrypt(key: dek, plaintext: bytes)
        return (enc.ciphertext, enc.iv, bytes.count)
    }

    public static func decryptFile(
        ciphertextB64: String,
        ivB64: String,
        dek: SymmetricKey
    ) throws -> String {
        let data = try aesGcmDecrypt(key: dek, ciphertextB64: ciphertextB64, ivB64: ivB64)
        return String(decoding: data, as: UTF8.self)
    }

    // MARK: - base64url (JWK fields)

    public static func base64urlEncode(_ data: Data) -> String {
        data.base64EncodedString()
            .replacingOccurrences(of: "+", with: "-")
            .replacingOccurrences(of: "/", with: "_")
            .replacingOccurrences(of: "=", with: "")
    }

    public static func base64urlDecode(_ s: String) -> Data? {
        var str = s
            .replacingOccurrences(of: "-", with: "+")
            .replacingOccurrences(of: "_", with: "/")
        while str.count % 4 != 0 { str += "=" }
        return Data(base64Encoded: str)
    }
}

File: RxCodeAndroid/app/src/main/java/app/rxlab/rxcode/crypto/SecretsCrypto.kt
package app.rxlab.rxcode.crypto

import app.rxlab.rxcode.proto.SecretsEnvironmentKey
import app.rxlab.rxcode.proto.SecretsUserKey
import org.bouncycastle.asn1.x9.ECNamedCurveTable
import org.bouncycastle.crypto.agreement.ECDHBasicAgreement
import org.bouncycastle.crypto.digests.SHA256Digest
import org.bouncycastle.crypto.generators.ECKeyPairGenerator
import org.bouncycastle.crypto.generators.HKDFBytesGenerator
import org.bouncycastle.crypto.params.ECDomainParameters
import org.bouncycastle.crypto.params.ECKeyGenerationParameters
import org.bouncycastle.crypto.params.ECPrivateKeyParameters
import org.bouncycastle.crypto.params.ECPublicKeyParameters
import org.bouncycastle.crypto.params.HKDFParameters
import org.json.JSONObject
import java.math.BigInteger
import java.security.SecureRandom
import java.util.Base64
import javax.crypto.Cipher
import javax.crypto.spec.GCMParameterSpec
import javax.crypto.spec.SecretKeySpec

/**
 * Byte-for-byte port of `RxCodeCore/Secrets/SecretsCrypto.swift` (itself a port
 * of github-pm's `lib/secrets/crypto.ts`). Keeping this interoperable with the
 * macOS/iOS/web implementations is the whole point: a secret encrypted on any
 * platform must decrypt here and vice-versa.
 *
 * Scheme:
 * - 32-byte **PRF output** comes from the user's passkey (see [SecretsKeyVault]).
 * - **KEK** = `HKDF-SHA256(prf, salt: ∅, info: ".../kek")`.
 * - Each user has an **ECDH P-256 keypair**; the private key is a JWK
 *   AES-GCM-wrapped under the KEK.
 * - Each environment has a random 32-byte **DEK**, wrapped under the owner's KEK
 *   (`wrapMode: "kek"`) or sealed to a public key (`wrapMode: "hpke"`).
 * - Files are AES-GCM encrypted under the DEK.
 *
 * AES-GCM layout matches WebCrypto: `iv` (12 bytes) stored separately, and
 * `ciphertext` is `rawCiphertext ‖ 16-byte tag` (JCA's GCM already appends the
 * tag to its output, matching WebCrypto).
 */
object SecretsCrypto {

    // Constants — must match github-pm/lib/secrets/constants.ts.
    val prfSalt: ByteArray = "github-pm-secrets-v1-prf-salt!!!".toByteArray(Charsets.UTF_8)
    private val hkdfInfoKEK = "github-pm-secrets-v1/kek".toByteArray(Charsets.UTF_8)
    private val hkdfInfoHPKE = "github-pm-secrets-v1/hpke".toByteArray(Charsets.UTF_8)
    const val WEBAUTHN_RP_ID = "rxlab.app"

    class CryptoException(message: String) : Exception(message)

    private val x9 = ECNamedCurveTable.getByName("P-256")
    private val domain = ECDomainParameters(x9.curve, x9.g, x9.n, x9.h)
    private val random = SecureRandom()

    private fun b64(data: ByteArray): String = Base64.getEncoder().encodeToString(data)
    private fun b64d(s: String): ByteArray = Base64.getDecoder().decode(s)

    // MARK: - KEK

    /** `HKDF-SHA256(prf, salt: ∅, info: ".../kek")` → 32-byte AES key. */
    fun deriveKEK(prfOutput: ByteArray): ByteArray =
        hkdf(prfOutput, ByteArray(0), hkdfInfoKEK, 32)

    private fun hkdf(ikm: ByteArray, salt: ByteArray, info: ByteArray, length: Int): ByteArray {
        val gen = HKDFBytesGenerator(SHA256Digest())
        gen.init(HKDFParameters(ikm, salt, info))
        val out = ByteArray(length)
        gen.generateBytes(out, 0, length)
        return out
    }

    // MARK: - AES-GCM (WebCrypto-compatible)

    /** Encrypts, returning base64 `ciphertext` (`raw ‖ tag`) + base64 `iv`. */
    fun aesGcmEncrypt(key: ByteArray, plaintext: ByteArray): Pair<String, String> {
        val iv = ByteArray(12).also { random.nextBytes(it) }
        val cipher = Cipher.getInstance("AES/GCM/NoPadding")
        cipher.init(Cipher.ENCRYPT_MODE, SecretKeySpec(key, "AES"), GCMParameterSpec(128, iv))
        val combined = cipher.doFinal(plaintext) // raw ‖ tag
        return b64(combined) to b64(iv)
    }

    fun aesGcmDecrypt(key: ByteArray, ciphertextB64: String, ivB64: String): ByteArray {
        val ctTag = b64d(ciphertextB64)
        val iv = b64d(ivB64)
        if (ctTag.size < 16) throw CryptoException("Ciphertext is too short to contain an auth tag.")
        val cipher = Cipher.getInstance("AES/GCM/NoPadding")
        cipher.init(Cipher.DECRYPT_MODE, SecretKeySpec(key, "AES"), GCMParameterSpec(128, iv))
        return cipher.doFinal(ctTag)
    }

    // MARK: - DEK

    fun generateDEK(): ByteArray = ByteArray(32).also { random.nextBytes(it) }

    // MARK: - ECDH P-256 / HPKE

    private fun bigIntTo32(value: BigInteger): ByteArray {
        val raw = value.toByteArray()
        return when {
            raw.size == 32 -> raw
            raw.size == 33 && raw[0] == 0.toByte() -> raw.copyOfRange(1, 33)
            raw.size < 32 -> ByteArray(32 - raw.size) + raw
            else -> raw.copyOfRange(raw.size - 32, raw.size)
        }
    }

    private fun importPublicKey(b64: String): ECPublicKeyParameters {
        val point = domain.curve.decodePoint(b64d(b64))
        return ECPublicKeyParameters(point, domain)
    }

    private fun ecdhSharedSecret(
        priv: ECPrivateKeyParameters,
        pub: ECPublicKeyParameters,
    ): ByteArray {
        val agree = ECDHBasicAgreement()
        agree.init(priv)
        return bigIntTo32(agree.calculateAgreement(pub)) // X coordinate, like WebCrypto/CryptoKit
    }

    private fun hpkeKey(shared: ByteArray): ByteArray = hkdf(shared, ByteArray(0), hkdfInfoHPKE, 32)

    fun sealToPublicKey(plaintext: ByteArray, recipientPublicKeyB64: String): Triple<String, String, String> {
        val recipient = importPublicKey(recipientPublicKeyB64)
        val gen = ECKeyPairGenerator().apply { init(ECKeyGenerationParameters(domain, random)) }
        val ephemeral = gen.generateKeyPair()
        val ephPriv = ephemeral.private as ECPrivateKeyParameters
        val ephPub = ephemeral.public as ECPublicKeyParameters
        val shared = ecdhSharedSecret(ephPriv, recipient)
        val enc = aesGcmEncrypt(hpkeKey(shared), plaintext)
        val ephB64 = b64(ephPub.q.getEncoded(false))
        return Triple(enc.first, enc.second, ephB64)
    }

    fun unsealWithPrivateKey(
        ciphertextB64: String,
        ivB64: String,
        ephemeralPublicKeyB64: String,
        recipientPrivateKey: ECPrivateKeyParameters,
    ): ByteArray {
        val ephPub = importPublicKey(ephemeralPublicKeyB64)
        val shared = ecdhSharedSecret(recipientPrivateKey, ephPub)
        return aesGcmDecrypt(hpkeKey(shared), ciphertextB64, ivB64)
    }

    // MARK: - User private key (JWK)

    private fun importPrivateKeyJWK(jwk: JSONObject): ECPrivateKeyParameters {
        val d = jwk.optString("d", "")
        if (d.isBlank()) throw CryptoException("Stored private key is not a valid P-256 JWK.")
        val dBytes = base64urlDecode(d) ?: throw CryptoException("Malformed JWK private scalar.")
        if (dBytes.size != 32) throw CryptoException("Stored private key is not a valid P-256 JWK.")
        return ECPrivateKeyParameters(BigInteger(1, dBytes), domain)
    }

    fun unwrapPrivateKey(wrappedPrivateKey: String, wrapIv: String, kek: ByteArray): ECPrivateKeyParameters {
        val json = aesGcmDecrypt(kek, wrappedPrivateKey, wrapIv)
        return importPrivateKeyJWK(JSONObject(String(json, Charsets.UTF_8)))
    }

    // MARK: - Flows (mirror SecretsFlows.swift)

    /** Fresh DEK wrapped under the owner KEK (`wrapMode: "kek"`). */
    fun buildOwnerEnvironmentKey(kek: ByteArray): Pair<app.rxlab.rxcode.proto.EnvironmentKeyBody, ByteArray> {
        val dek = generateDEK()
        val wrap = aesGcmEncrypt(kek, dek)
        val body = app.rxlab.rxcode.proto.EnvironmentKeyBody(
            wrapMode = "kek",
            wrappedDek = wrap.first,
            wrapIv = wrap.second,
            ephemeralPublicKey = null,
        )
        return body to dek
    }

    /** Resolves an environment key row into the raw DEK bytes. */
    fun unwrapDEK(
        envKey: SecretsEnvironmentKey,
        kek: ByteArray,
        userPrivateKey: ECPrivateKeyParameters?,
    ): ByteArray = when (envKey.wrapMode) {
        "kek" -> {
            val iv = envKey.wrapIv ?: throw CryptoException("Wrapped key is missing its IV.")
            aesGcmDecrypt(kek, envKey.wrappedDek, iv)
        }
        "hpke" -> {
            val priv = userPrivateKey
                ?: throw CryptoException("This environment was shared with you; your enrollment key is required to open it.")
            val iv = envKey.wrapIv
            val eph = envKey.ephemeralPublicKey
            if (iv == null || eph == null) throw CryptoException("Shared key is missing its ephemeral public key or IV.")
            unsealWithPrivateKey(envKey.wrappedDek, iv, eph, priv)
        }
        else -> throw CryptoException("Unsupported wrap mode: ${envKey.wrapMode}.")
    }

    fun unwrapUserPrivateKey(userKey: SecretsUserKey, kek: ByteArray): ECPrivateKeyParameters? {
        val wrapped = userKey.wrappedPrivateKey ?: return null
        val iv = userKey.wrapIv ?: return null
        return unwrapPrivateKey(wrapped, iv, kek)
    }

    // MARK: - File helpers

    fun encryptFile(plaintext: String, dek: ByteArray): Triple<String, String, Int> {
        val bytes = plaintext.toByteArray(Charsets.UTF_8)
        val enc = aesGcmEncrypt(dek, bytes)
        return Triple(enc.first, enc.second, bytes.size)
    }

    fun decryptFile(ciphertextB64: String, ivB64: String, dek: ByteArray): String =
        String(aesGcmDecrypt(dek, ciphertextB64, ivB64), Charsets.UTF_8)

    // MARK: - base64url

    private fun base64urlDecode(s: String): ByteArray? {
        var str = s.replace('-', '+').replace('_', '/')
        while (str.length % 4 != 0) str += "="
        return runCatching { Base64.getDecoder().decode(str) }.getOrNull()
    }
}
